File: stock-service/src/main/java/rs/raf/stock_service/security/JwtAuthenticationFilter.java
package rs.raf.stock_service.security;

import io.jsonwebtoken.Claims;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import rs.raf.stock_service.utils.JwtTokenUtil;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.List;

/*
        Class is called on every Http request/response to check validity of token.
        It is set in the security config as before filter for every call.
     */
@Component
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private final JwtTokenUtil jwtTokenUtil;

    public JwtAuthenticationFilter(JwtTokenUtil jwtTokenUtil) {
        this.jwtTokenUtil = jwtTokenUtil;
    }


    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {
        String token = getJwtFromRequest(request);

        if (token != null && jwtTokenUtil.validateToken(token)) {
            Claims claims = jwtTokenUtil.getClaimsFromToken(token);
            String email = claims.getSubject();

            String role = claims.get("role", String.class);
            GrantedAuthority authority = new SimpleGrantedAuthority("ROLE_" + role.toUpperCase());
            List<GrantedAuthority> authorities = List.of(authority);

            UserDetails userDetails = new org.springframework.security.core.userdetails.User(email, "", authorities);

            UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(
                    userDetails, null, authorities);
            SecurityContextHolder.getContext().setAuthentication(authentication);
        }

        filterChain.doFilter(request, response);
    }


    private String getJwtFromRequest(HttpServletRequest request) {
        String bearerToken = request.getHeader("Authorization");
        System.out.println(bearerToken);
        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
            return bearerToken.substring(7);
        }
        return null;
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/configuration/JwtAuthenticationFilter.kt
package pack.userservicekotlin.configuration

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import pack.userservicekotlin.utils.JwtTokenUtil
import java.io.IOException
import java.util.*
import java.util.List
import javax.servlet.ServletException

/*
       Class is called on every Http request/response to check validity of token.
       It is set in the security config as before filter for every call.
 */
@Component
class JwtAuthenticationFilter(
    private val jwtTokenUtil: JwtTokenUtil,
) : OncePerRequestFilter() {
    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val token = getJwtFromRequest(request)

        if (token != null && jwtTokenUtil.validateToken(token)) {
            val claims = jwtTokenUtil.getClaimsFromToken(token)
            val email = claims.subject

            val role = claims.get("role", String::class.java)
            val authority: GrantedAuthority = SimpleGrantedAuthority("ROLE_" + role.uppercase(Locale.getDefault()))
            val authorities = List.of(authority)

            val userDetails: UserDetails = User(email, "", authorities)

            val authentication =
                UsernamePasswordAuthenticationToken(
                    userDetails,
                    null,
                    authorities,
                )
            SecurityContextHolder.getContext().authentication = authentication
        }

        filterChain.doFilter(request, response)
    }

    private fun getJwtFromRequest(request: HttpServletRequest): String? {
        val bearerToken = request.getHeader("Authorization")
        println(bearerToken)
        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
            return bearerToken.substring(7)
        }
        return null
    }
}
